gobblin-service/src/main/java/org/apache/gobblin/service/modules/template/FlowTemplate.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.template;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+import com.typesafe.config.Config;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.gobblin.annotation.Alpha;
+import org.apache.gobblin.runtime.api.JobTemplate;
+import org.apache.gobblin.runtime.api.Spec;
+import org.apache.gobblin.service.modules.dataset.DatasetDescriptor;
+import org.apache.gobblin.service.modules.flowgraph.Dag;
+
+/**
+ * An interface primarily for representing a flow of {@link JobTemplate}s. It also has
+ * method for retrieving required configs for every {@link JobTemplate} in the flow.
+ */
+@Alpha
+public interface FlowTemplate extends Spec {
+
+  /**
+   * @return the {@link Collection} of {@link JobTemplate}s that belong to this {@link FlowTemplate}.
+   */
+  List<JobTemplate> getJobTemplates();
+
+  /**
+   *
+   * @return the {@link Dag<JobTemplate>} that backs the {@link FlowTemplate}.
+   */
+  Dag<JobTemplate> getDag() throws IOException;
+
+  /**
+   *
+   * @return all configuration inside pre-written template.
+   */
+  Config getRawTemplateConfig();
+
+  /**
+   * @return list of input/output {@link DatasetDescriptor}s for the {@link FlowTemplate}.
+   */
+  List<Pair<DatasetDescriptor, DatasetDescriptor>> getInputOutputDatasetDescriptors();
```

Comment:
We've talked about this in meeting. Most of the interface defined are exposing its properties. For this one, the actual use case is check if a give pair of input and ouput `DatasetDescriptor`s is acceptable in the flow template. We could just expose `isAcceptable(Pair<DatasetDescriptor, DatasetDescriptor>)` to encapsulate the object more.

Reply:
Not quite. A FlowEdge may accept a set of I/O dataset descriptors. During path finding, we need to potentially look at each pair in this set to determine if there is at least one I/O descriptor that is compatible with the target. Hence the need for this API.